File: assignment4/java/Scheduler.java
import java.util.*;

/** Employee Shift Scheduler -- Java implementation. */
public class Scheduler {

    static final String[] DAYS = {
        "Monday", "Tuesday", "Wednesday", "Thursday",
        "Friday", "Saturday", "Sunday"
    };
    static final String[] SHIFTS = {"morning", "afternoon", "evening"};

    static final int MIN_PER_SHIFT = 2;
    static final int MAX_PER_SHIFT = 3;
    static final int MAX_DAYS_PER_EMPLOYEE = 5;

    static Map<String, Map<String, String>> employees = buildEmployees();
    static Map<String, Map<String, List<String>>> schedule = new LinkedHashMap<>();
    static Map<String, Integer> daysWorked = new HashMap<>();
    static List<String> notes = new ArrayList<>();
    static Random rng = new Random(42);

    public static void main(String[] args) {
        initializeSchedule();
        for (String emp : employees.keySet()) {
            daysWorked.put(emp, 0);
        }

        // Pass 1: honor preferences
        for (Map.Entry<String, Map<String, String>> entry : employees.entrySet()) {
            String emp = entry.getKey();
            for (Map.Entry<String, String> pref : entry.getValue().entrySet()) {
                assignWithConflictResolution(emp, pref.getKey(), pref.getValue());
            }
        }

        // Pass 2: enforce minimum staffing
        fillMinimumStaffing();

        printSchedule();
    }

    static void initializeSchedule() {
        for (String day : DAYS) {
            Map<String, List<String>> shifts = new LinkedHashMap<>();
            for (String shift : SHIFTS) {
                shifts.put(shift, new ArrayList<>());
            }
            schedule.put(day, shifts);
        }
    }

    static boolean isShiftFull(String day, String shift) {
        return schedule.get(day).get(shift).size() >= MAX_PER_SHIFT;
    }

    static boolean workedThatDay(String employee, String day) {
        for (String shift : SHIFTS) {
            if (schedule.get(day).get(shift).contains(employee)) {
                return true;
            }
        }
        return false;
    }

    static boolean tryAssign(String employee, String day, String shift) {
        if (daysWorked.get(employee) >= MAX_DAYS_PER_EMPLOYEE) return false;
        if (workedThatDay(employee, day)) return false;
        if (isShiftFull(day, shift)) return false;
        schedule.get(day).get(shift).add(employee);
        daysWorked.merge(employee, 1, Integer::sum);
        return true;
    }

    static void assignWithConflictResolution(String employee, String day, String preferredShift) {
        if (tryAssign(employee, day, preferredShift)) {
            return;
        }

        for (String alt : SHIFTS) {
            if (alt.equals(preferredShift)) continue;
            if (tryAssign(employee, day, alt)) {
                notes.add(employee + ": conflict on " + day + " " + preferredShift
                        + " -> reassigned to " + day + " " + alt + ".");
                return;
            }
        }

        int idx = Arrays.asList(DAYS).indexOf(day);
        String nextDay = DAYS[(idx + 1) % DAYS.length];
        if (tryAssign(employee, nextDay, preferredShift)) {
            notes.add(employee + ": conflict on " + day + " " + preferredShift
                    + " -> reassigned to " + nextDay + " " + preferredShift + ".");
            return;
        }

        notes.add(employee + ": could not honor " + day + " " + preferredShift
                + " (no available shifts).");
    }

    static void fillMinimumStaffing() {
        List<String> empList = new ArrayList<>(employees.keySet());
        for (String day : DAYS) {
            for (String shift : SHIFTS) {
                while (schedule.get(day).get(shift).size() < MIN_PER_SHIFT) {
                    List<String> candidates = new ArrayList<>();
                    for (String e : empList) {
                        if (daysWorked.get(e) < MAX_DAYS_PER_EMPLOYEE
                                && !workedThatDay(e, day)) {
                            candidates.add(e);
                        }
                    }
                    if (candidates.isEmpty()) {
                        System.out.println("  WARNING: cannot meet minimum on "
                                + day + " " + shift);
                        break;
                    }
                    String chosen = candidates.get(rng.nextInt(candidates.size()));
                    schedule.get(day).get(shift).add(chosen);
                    daysWorked.merge(chosen, 1, Integer::sum);
                }
            }
        }
    }

    static void printSchedule() {
        String bar = repeat("=", 80);
        System.out.println(bar);
        String title = " Weekly Employee Schedule ";
        int pad = (80 - title.length()) / 2;
        System.out.println(repeat("=", pad) + title + repeat("=", 80 - pad - title.length()));
        System.out.println(bar);

        int colWidth = 22;
        StringBuilder header = new StringBuilder(padRight("Day", 10));
        for (String s : SHIFTS) {
            header.append(padRight(capitalize(s), colWidth));
        }
        System.out.println(header);
        System.out.println(repeat("-", 80));

        for (String day : DAYS) {
            StringBuilder row = new StringBuilder(padRight(day, 10));
            for (String shift : SHIFTS) {
                List<String> people = schedule.get(day).get(shift);
                String cell = people.isEmpty() ? "(empty)" : String.join(", ", people);
                row.append(padRight(cell, colWidth));
            }
            System.out.println(row);
        }
        System.out.println();
        System.out.println("Days worked per employee:");
        List<String> emps = new ArrayList<>(daysWorked.keySet());
        Collections.sort(emps);
        for (String e : emps) {
            System.out.println("  " + padRight(e, 8) + " " + daysWorked.get(e));
        }

        if (!notes.isEmpty()) {
            System.out.println();
            System.out.println("Conflict resolution notes:");
            for (String n : notes) {
                System.out.println("  - " + n);
            }
        }
    }

    static String padRight(String s, int width) {
        if (s.length() >= width) return s.substring(0, width - 1) + " ";
        StringBuilder sb = new StringBuilder(s);
        while (sb.length() < width) sb.append(' ');
        return sb.toString();
    }

    static String repeat(String s, int n) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) sb.append(s);
        return sb.toString();
    }

    static String capitalize(String s) {
        return Character.toUpperCase(s.charAt(0)) + s.substring(1);
    }

    static Map<String, Map<String, String>> buildEmployees() {
        Map<String, Map<String, String>> m = new LinkedHashMap<>();
        m.put("Alice", prefs(
                "Monday", "morning",
                "Tuesday", "morning",
                "Wednesday", "morning",
                "Thursday", "afternoon",
                "Friday", "morning"));
        m.put("Bob", prefs(
                "Monday", "afternoon",
                "Tuesday", "afternoon",
                "Wednesday", "evening",
                "Thursday", "evening",
                "Friday", "afternoon"));
        m.put("Carol", prefs(
                "Monday", "evening",
                "Tuesday", "evening",
                "Wednesday", "evening",
                "Saturday", "morning",
                "Sunday", "morning"));
        m.put("David", prefs(
                "Monday", "morning",
                "Tuesday", "morning",
                "Thursday", "morning",
                "Friday", "morning",
                "Saturday", "afternoon"));
        m.put("Eve", prefs(
                "Wednesday", "afternoon",
                "Thursday", "afternoon",
                "Friday", "afternoon",
                "Saturday", "evening",
                "Sunday", "evening"));
        m.put("Frank", prefs(
                "Monday", "morning",
                "Wednesday", "morning",
                "Friday", "evening",
                "Saturday", "morning",
                "Sunday", "afternoon"));
        m.put("Grace", prefs(
                "Tuesday", "evening",
                "Wednesday", "evening",
                "Thursday", "morning",
                "Saturday", "evening",
                "Sunday", "evening"));
        m.put("Henry", prefs(
                "Monday", "evening",
                "Tuesday", "morning",
                "Thursday", "evening",
                "Friday", "morning",
                "Sunday", "morning"));
        m.put("Iris", prefs(
                "Monday", "morning",
                "Wednesday", "afternoon",
                "Thursday", "afternoon",
                "Friday", "evening",
                "Saturday", "afternoon"));
        m.put("Jack", prefs(
                "Tuesday", "afternoon",
                "Wednesday", "morning",
                "Thursday", "evening",
                "Saturday", "morning",
                "Sunday", "afternoon"));
        return m;
    }

    static Map<String, String> prefs(String... pairs) {
        Map<String, String> p = new LinkedHashMap<>();
        for (int i = 0; i < pairs.length; i += 2) {
            p.put(pairs[i], pairs[i + 1]);
        }
        return p;
    }
}

File: assignment4/python/scheduler.py
"""Employee Shift Scheduler -- Python implementation."""

import random
from collections import defaultdict

DAYS = [
    "Monday", "Tuesday", "Wednesday", "Thursday",
    "Friday", "Saturday", "Sunday",
]
SHIFTS = ["morning", "afternoon", "evening"]

MIN_PER_SHIFT = 2
MAX_PER_SHIFT = 3
MAX_DAYS_PER_EMPLOYEE = 5

EMPLOYEES = {
    "Alice":   {"Monday": "morning",   "Tuesday": "morning",    "Wednesday": "morning",   "Thursday": "afternoon", "Friday": "morning"},
    "Bob":     {"Monday": "afternoon", "Tuesday": "afternoon",  "Wednesday": "evening",   "Thursday": "evening",   "Friday": "afternoon"},
    "Carol":   {"Monday": "evening",   "Tuesday": "evening",    "Wednesday": "evening",   "Saturday": "morning",   "Sunday": "morning"},
    "David":   {"Monday": "morning",   "Tuesday": "morning",    "Thursday": "morning",    "Friday": "morning",     "Saturday": "afternoon"},
    "Eve":     {"Wednesday": "afternoon", "Thursday": "afternoon", "Friday": "afternoon", "Saturday": "evening",   "Sunday": "evening"},
    "Frank":   {"Monday": "morning",   "Wednesday": "morning",  "Friday": "evening",      "Saturday": "morning",   "Sunday": "afternoon"},
    "Grace":   {"Tuesday": "evening",  "Wednesday": "evening",  "Thursday": "morning",    "Saturday": "evening",   "Sunday": "evening"},
    "Henry":   {"Monday": "evening",   "Tuesday": "morning",    "Thursday": "evening",    "Friday": "morning",     "Sunday": "morning"},
    "Iris":    {"Monday": "morning",   "Wednesday": "afternoon", "Thursday": "afternoon", "Friday": "evening",     "Saturday": "afternoon"},
    "Jack":    {"Tuesday": "afternoon", "Wednesday": "morning", "Thursday": "evening",    "Saturday": "morning",   "Sunday": "afternoon"},
}


def initialize_schedule():
    return {day: {shift: [] for shift in SHIFTS} for day in DAYS}


def is_shift_full(schedule, day, shift):
    return len(schedule[day][shift]) >= MAX_PER_SHIFT


def employee_worked_that_day(schedule, employee, day):
    for shift in SHIFTS:
        if employee in schedule[day][shift]:
            return True
    return False


def try_assign(schedule, days_worked, employee, day, shift):
    if days_worked[employee] >= MAX_DAYS_PER_EMPLOYEE:
        return False
    if employee_worked_that_day(schedule, employee, day):
        return False
    if is_shift_full(schedule, day, shift):
        return False
    schedule[day][shift].append(employee)
    days_worked[employee] += 1
    return True


def assign_with_conflict_resolution(schedule, days_worked, employee, day, preferred_shift):
    if try_assign(schedule, days_worked, employee, day, preferred_shift):
        return preferred_shift, day, "preferred"

    for alt in SHIFTS:
        if alt == preferred_shift:
            continue
        if try_assign(schedule, days_worked, employee, day, alt):
            return alt, day, "same-day-alternate"

    next_day = DAYS[(DAYS.index(day) + 1) % len(DAYS)]
    if try_assign(schedule, days_worked, employee, next_day, preferred_shift):
        return preferred_shift, next_day, "next-day"

    return None, None, "unassigned"


def fill_minimum_staffing(schedule, days_worked):
    employees = list(EMPLOYEES.keys())
    for day in DAYS:
        for shift in SHIFTS:
            while len(schedule[day][shift]) < MIN_PER_SHIFT:
                candidates = [
                    e for e in employees
                    if days_worked[e] < MAX_DAYS_PER_EMPLOYEE
                    and not employee_worked_that_day(schedule, e, day)
                ]
                if not candidates:
                    print(f"  WARNING: cannot meet minimum on {day} {shift}")
                    break
                chosen = random.choice(candidates)
                schedule[day][shift].append(chosen)
                days_worked[chosen] += 1


def build_schedule():
    schedule = initialize_schedule()
    days_worked = defaultdict(int)
    notes = []

    # Pass 1: honor preferences
    for employee, prefs in EMPLOYEES.items():
        for day, preferred_shift in prefs.items():
            assigned_shift, assigned_day, status = assign_with_conflict_resolution(
                schedule, days_worked, employee, day, preferred_shift)
            if status == "unassigned":
                notes.append(
                    f"{employee}: could not honor {day} {preferred_shift} "
                    f"(no available shifts).")
            elif status != "preferred":
                notes.append(
                    f"{employee}: conflict on {day} {preferred_shift} -> "
                    f"reassigned to {assigned_day} {assigned_shift}.")

    # Pass 2: enforce minimum staffing
    fill_minimum_staffing(schedule, days_worked)
    return schedule, days_worked, notes


def print_schedule(schedule, days_worked, notes):
    bar = "=" * 80
    print(bar)
    print(" Weekly Employee Schedule ".center(80, "="))
    print(bar)

    col_width = 22
    header = "Day".ljust(10) + "".join(s.capitalize().ljust(col_width) for s in SHIFTS)
    print(header)
    print("-" * 80)
    for day in DAYS:
        cells = []
        for shift in SHIFTS:
            people = schedule[day][shift]
            cells.append(", ".join(people) if people else "(empty)")
        print(day.ljust(10) + "".join(c.ljust(col_width) for c in cells))

    print()
    print("Days worked per employee:")
    for emp in sorted(days_worked.keys()):
        print(f"  {emp:<8} {days_worked[emp]}")

    if notes:
        print()
        print("Conflict resolution notes:")
        for note in notes:
            print(f"  - {note}")


def main():
    random.seed(42)
    schedule, days_worked, notes = build_schedule()
    print_schedule(schedule, days_worked, notes)


if __name__ == "__main__":
    main()
